src/test/java/org/mule/test/http/functional/HttpListenerOverloadTestCase.java

```diff
@@ -0,0 +1,147 @@
+import java.util.stream.Collectors;
+
+import io.qameta.allure.Feature;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.fluent.Executor;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+@Feature(HTTP_EXTENSION)
+public class HttpListenerOverloadTestCase extends AbstractHttpTestCase {
+
+  // This must match the default pool size in ContainerThreadPoolsConfig
+  private static final int MAX_CONNECTIONS = 256;
+
+  @Rule
+  public DynamicPort listenPort = new DynamicPort("port");
+
+  private CountDownLatch activeProcessorsStarted;
+  private CountDownLatch activeProcessorsFinished;
+  private Latch keepProcessorsActive = new Latch();
+  private AtomicInteger numberOfRequest = new AtomicInteger();
+  private Executor httpClientExecutor;
+  private ConcurrentLinkedQueue<Throwable> accumulatedErrors = new ConcurrentLinkedQueue<>();
+
+  @Override
+  protected String getConfigFile() {
+    return "http-listener-overload-config.xml";
+  }
+
+  @Before
+  public void setup() {
+    // Need to configure the maximum number of connections of HttpClient, because the default is less than
+    // the default number of workers in the HTTP listener.
+    PoolingHttpClientConnectionManager mgr = new PoolingHttpClientConnectionManager();
+    mgr.setDefaultMaxPerRoute(MAX_CONNECTIONS);
+    mgr.setMaxTotal(MAX_CONNECTIONS);
+    httpClientExecutor = Executor.newInstance(HttpClientBuilder.create().setConnectionManager(mgr).build());
+  }
+
+  /**
+   * Phase 1: Start N requests to the listener, which will use all workers from the pool to wait on a latch inside a message processor.
+   * Phase 2: Start a few requests to the listener, checking that they return "server busy" status code without getting into the flow message processors.
+   * Phase 3: Release the latch, allowing the N requests to continue and verifying they succeed.
+   * Phase 4: Verify that all requests were processed and no errors were thrown.
+   */
+  @Test
+  public void overloadScenario() throws Exception {
+    final String url = format("http://localhost:%s/", listenPort.getNumber());
+
+    try {
+      sendRequestUntilNoMoreWorkers("testFlow", url, MAX_CONNECTIONS - 1);
+
+      for (int i = 0; i < 5; i++) {
+        final HttpResponse response = Request.Get(url).connectTimeout(RECEIVE_TIMEOUT).execute().returnResponse();
+        assertThat(response.getStatusLine().getStatusCode(), is(SERVICE_UNAVAILABLE.getStatusCode()));
+        assertThat(response.getStatusLine().getReasonPhrase(), is(SERVICE_UNAVAILABLE.getReasonPhrase()));
+        assertThat(IOUtils.toString(response.getEntity().getContent()), is("Scheduler unavailable"));
+      }
+
+    } finally {
+      keepProcessorsActive.release();
+      if (!activeProcessorsFinished.await(15, TimeUnit.SECONDS)) {
```

Comment:
That a very large timeout, have you validated whether it can be lowered? You could use the flakiness detector runner to check.